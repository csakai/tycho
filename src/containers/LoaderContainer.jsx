import React from 'react';
import {DefaultLoadingManager} from 'three';
import {connect} from 'react-redux';
import * as AnimationActions from '../actions/AnimationActions';
import * as LoaderActions from '../actions/LoaderActions';
import ReduxService from '../services/ReduxService';
import SplashScreen from '../components/SplashScreen';

export class LoaderContainer extends React.Component {

  componentWillMount = () => {
    this.setState({hasEntered: false});
    DefaultLoadingManager.onProgress = this.onProgress;
  }

  /**
   * Updates the redux store with current loader status.
   *
   * @param {String} url - loaded url
   * @param {Number} count - number of textures loaded
   * @param {Number} total - total number of textures enqueued
   */
  onProgress = (url, count, total) => {
    this.props.action.setPercentLoaded(count, total);
    this.props.action.setTextureLoaded(url);
  }

  /**
   * Starts playing the scene.
   */
  enterScene = () => {
    this.props.action.setPlaying(true);
    this.setState({hasEntered: true});
  }

  render() {
    return (
      <SplashScreen
        percent={this.props.percent || 100}
        show={!this.state.hasEntered}
        enterScene={this.enterScene}
        pageText={this.props.pageText}
      />
    );
  }
}

export default connect(
  ReduxService.mapStateToProps(
    'loader.url',
    'loader.percent',
    'animation.playing',
    'data.pageText'
  ),
  ReduxService.mapDispatchToProps(
    AnimationActions,
    LoaderActions
  )
)(LoaderContainer);
