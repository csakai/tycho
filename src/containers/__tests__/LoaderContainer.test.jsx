import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {DefaultLoadingManager} from 'three';
import {LoaderContainer} from '../LoaderContainer';

describe('Loader Container', () => {
  let component, loaderContainer;

  beforeEach(() => {
    component = shallow(<LoaderContainer
      action={{
        setPercentLoaded: jest.fn(),
        setTextureLoaded: jest.fn(),
        setUserEntered: jest.fn()
      }}
    />);
    loaderContainer = component.instance();
  });

  describe('componentDidMount()', () => {
    it('should assign `onProgress` to the DefaultLoadingManager', () => {
      expect(DefaultLoadingManager.onProgress).toEqual(loaderContainer.onProgress);
    });
  });

  describe('onProgress()', () => {
    const url = 'myImage.jpg';
    const count = 5;
    const total = 8;
    
    it('should call the setPercentLoaded() action', () => {
      const spy = jest.spyOn(loaderContainer.props.action, 'setPercentLoaded');

      loaderContainer.onProgress(url, count, total);

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(count, total);
    });

    it('should call the setTextureLoaded() action', () => {
      const spy = jest.spyOn(loaderContainer.props.action, 'setTextureLoaded');

      loaderContainer.onProgress(url, count, total);

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(url);
    });
  });

  describe('enterScene()', () => {
    it('should call the setUserEntered() action with true', () => {
      const spy = jest.spyOn(loaderContainer.props.action, 'setUserEntered');

      loaderContainer.enterScene();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(true);
    });
  });

  describe('render()', () => {
    it('should render the loader container successfully', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
