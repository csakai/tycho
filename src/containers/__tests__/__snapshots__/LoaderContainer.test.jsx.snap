// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader Container render() should render the loader container successfully 1`] = `
<SplashScreen
  enterScene={[Function]}
  percent={0}
  show={true}
/>
`;
